gitindex/index.go

```diff
@@ -532,6 +539,229 @@ func newIgnoreMatcher(tree *object.Tree) (*ignore.Matcher, error) {
+	// discover what commits we indexed during our last build
+
+	existingRepository, err := options.BuildOptions.RepositoryMetadata()
+	if err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("failed to get repository metadata: %w", err)
+	}
+
+	if existingRepository == nil {
+		return nil, nil, nil, nil, fmt.Errorf("no existing shards found for repository")
+	}
+
+	// Check to see if the branch set is consistent with what we last indexed.
+	// If it isn't consistent, that we can't proceed with a delta build (and the caller should fall back to a
+	// normal one).
+
+	indexState := build.CompareBranches(existingRepository.Branches, options.BuildOptions.RepositoryDescription.Branches)
+	if indexState != build.IndexStateBranchVersion {
+		var existingBranchNames []string
+		for _, b := range existingRepository.Branches {
+			existingBranchNames = append(existingBranchNames, b.Name)
+		}
+
+		var optionsBranchNames []string
+		for _, b := range options.BuildOptions.RepositoryDescription.Branches {
+			optionsBranchNames = append(optionsBranchNames, b.Name)
+		}
+
+		existingBranchList := strings.Join(existingBranchNames, ", ")
+		optionsBranchList := strings.Join(optionsBranchNames, ", ")
+
+		return nil, nil, nil, nil, fmt.Errorf("requested branch set in build options (%q) != branch set found on disk (%q) - branch set must be the same for delta shards", optionsBranchList, existingBranchList)
+	}
+
+	// branch => (path, sha1) => repo.
+	repos = map[fileKey]BlobLocation{}
+
+	// fileKey => branches
+	branchMap = map[fileKey][]string{}
+
+	// Branch => Repo => SHA1
+	//branchVersions := map[string]map[string]plumbing.Hash{}
+
+	// branch name -> git worktree at most current commit
+	branchToCurrentTree := make(map[string]*object.Tree, len(options.Branches))
+
+	for _, b := range options.Branches {
+		commit, err := getCommit(repository, options.BranchPrefix, b)
+		if err != nil {
+			return nil, nil, nil, nil, fmt.Errorf("getting last current commit for branch %q: %w", b, err)
+		}
+
+		tree, err := commit.Tree()
+		if err != nil {
+			return nil, nil, nil, nil, fmt.Errorf("getting current git tree for branch %q: %w", b, err)
+		}
+
+		branchToCurrentTree[b] = tree
+	}
+
+	rawURL := options.BuildOptions.RepositoryDescription.URL
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("parsing repository URL %q: %w", rawURL, err)
+	}
+
+	// TODO: Support repository submodules for delta builds
+	// For this prototype, we are ignoring repository submodules, which means that we can use the same
+	// blob location for all files
+	hackSharedBlobLocation := BlobLocation{
+		Repo: repository,
+		URL:  u,
+	}
+
+	// loop over all branches, calculate the diff between our
+	// last indexed commit and the current commit, and add files mentioned in the diff
+	for _, branch := range existingRepository.Branches {
+		lastIndexedCommit, err := getCommit(repository, "", branch.Version)
```

Comment:
This isn't necessary if we specifically want to lookup a _commit_ instead of a branch name (notice that we're passing `branch.Version` to the function. 

This comment implies that this function is intended to fetch objects for both branches and commits: 

https://github.com/sourcegraph/zoekt/blob/86e6a545a7d47b35b4a4d7d1fad956ef43a592f4/gitindex/index.go#L158

Is this not the case?

Reply:
I don't know :) I was just comparing the two implementations and looking for inconsistencies.

Reply:
Can we resolve this comment chain?